Produce Convexity Check Warning Only When Appropriate

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #10299 
 - The convexity check is only supposed to throw a warning message when the Polygon Clipping Algorithm is employed with the ConvexWeilerAtherton algorithm for shading.  Previously, this was being produced for Pixel Counting and other shading algorithms inappropriately.  The code was modified to match the intent of the documentation and the warning messages being produced by the code.  The solution required that the get routine for the Shadowing Calculation input be separated from some of the processing that is done and the get routine had to be moved up in the order of simulation.  After the fix, the only differences in the results are a change in order of a couple of lines in the EIO file (change in order, not in values in the report).  All other output is identical to before the fix as anticipated.  Six new unit tests look at various combinations of input to make sure things are being read in and set properly.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
10299 Split GetShadowingInput

This commit splits GetShadowingInput into two separate routines and will test to see if the split has been done correctly.  The split is important because the actual getting of the input needs to be moved up in the order of calls to resolve this defect, but some of the processing needs to stay where it is at because it needs zones and surfaces read in.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Separation of Get and Process

The previous version of GetShadowingInput obtained the ShadowCalculation input but then also did some processing.  In order to solve #10299, these needed to be separated so that when surfaces are read in that it knows the ShadowCalculation input and thus can produce error messages appropriately.  This commit separates the two parts of the previous routine into two routines which are now called at the appropriate places.  The next step will be to implement the solution and create a unit test.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Fix and Reverse Old Unit

Fix to the problem.  Testing reveals that the code now does what it is supposed to do.  Also, reversed some changes to an existing unit test as the changes were causing some issues in the test suite.  Hopefully this makes those issues disappear.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Unit Test and Fixes

Addition of new unit tests and fixes to the one that doesn't work properly for Windows (experimental solution based on surrounding code).  If this works, this is a potential pull request version.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Unit Test Change Backup

Backed out something that was tried but resulted in more issues.

RKStrand committed:
10299 Correction of Unit Test enum

Correction of enum comparisons to eliminate errors in energyplus_tests

RKStrand committed:
10299 Move a PixelCounting Check

An error message was being skipped because of the movement of the Get part of the old routine.  anyScheduledShadingSurface was not set yet so the error message was never produced.  This change should fix this problem and eliminate the change in the .err file for SolarShadingTestGPU.idf.

RKStrand committed:
10299 Correction of Unit Test Errors

A couple of issues were seen in ci results.  This commit is an attempt to fix them.

RKStrand committed:
10299 Another Round of Unit Test Fixes

Last round was closer to getting rid of all of the unit test fails.  Hopefully this gets everything fixed up and ready to go.

RKStrand committed:
10299 Get Rid of Unneeded Get Flag

A get flag was added early on during work on this defect.  It is not necessary and may be causing some issues with API runs.  Even if that isn't the problem, there is no reason to keep this flag/code--it's unnecessary.

RKStrand committed:
10299 Clang Format Final

Missed clang formatting on a file...

mjwitte commented on src/EnergyPlus/DataSystemVariables.hh:
```diff
@@ -112,6 +112,11 @@ struct SystemVarsData : BaseGlobalStruct
+    bool DisableSelfShadingBetweenGroup = false;
+
+    int shadingGroupsNum = 0; // number of shading groups
+    Array1D_string zoneName;  // array of zone names in user input
```
Seems odd that all of these shading variables are in `DataSystemVariables` instead of `SolarShading`. I'm not suggesting you have to move these in this PR, but 'zoneName' needs to be more specific, like `shadingGroupZoneListNames`.

Also, if you make this a `std::vector` and push to it when reading the input, then you can use it's size and drop `shadingGroupsNum`.

RKStrand replied:
Yeah, I don't know why these are in DataSystemVariables and not SolarShading.  I guess I was looking to solve the problem and wasn't too concerned about the location of the variables (thought there must have been some reason--probably not a great assumption).

So, just to clarify and get this to move forward...you would like to see the zoneName changed to shadingGroupZoneListNames.  You would also like to see this converted to a vector as part of this work?

mjwitte replied:
> So, just to clarify and get this to move forward...you would like to see the zoneName changed to shadingGroupZoneListNames. 

Yes, please.

> You would also like to see this converted to a vector as part of this work?

That's up to you for this PR.

RKStrand replied:
Variable name change made in the latest commit.  Hopefully the ci comes back with all green as before.

RKStrand replied:
Ok, after some back and forth and some weird goings on yesterday, the variable name change has been made and ci is coming back like it should with only the EIO changes (only changes in the info location--values the same).  All good?

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -797,6 +717,92 @@ void GetShadowingInput(EnergyPlusData &state)
+                if (Found != 0) DisableSelfShadingGroups(i) = Found;
+            }
+
+            for (int SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; SurfNum++) {
```
I know this is existing code, and it's not a frequently used option, but this could be so much more efficient. Instead of looping through all of the zonegroups for each surface, this could simply go through the zones in each zonelist and loop  over the exterior surrfaces in each zone (`Zone.vZoneHTSurfaceList`). Mostly just noting this here, since this seemingly simple fix has already had a lot of scope creep.

RKStrand replied:
I agree that this would be much more efficient.  I'd like to have us make a mental note of this and deal with it at a later date if you are okay with that.  I wonder if there are more places where we could be much more efficient and we don't even realize it.  If only we all had the time to do that sort of efficiency review in the code...

RKStrand committed:
10299 Request to Improve Variable Name

Addressed review request to change the name of a variable to be more descriptive.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Back Out Variable Name Change

Something not quite right about this simply variable name change.  Things went from expected diffs only to "unexpected" diffs.

RKStrand committed:
Merge branch 'develop' into 10299CheckConvexityChangeCheckForShadowCalculations

RKStrand committed:
10299 Attempt 2 at Variable Rename

Trying again to rename a variable.  This seemed to cause a lot of issues yesterday, but was that because of changes to develop that were happening at the same time?

Myoldmopar reviewed (approved):
I'm fine with holding off on the efficiency refactor that was brought up by @mjwitte.  Let's just try to get this merged if it's ready.  I'll test it out.

Myoldmopar commented:
Yeah, it's all testing fine, and CI is all happy.  I'll merge this one.  Thanks @RKStrand
